### PR TITLE
node driver API refactoring

### PR DESCRIPTION
**Pull request by futuretea**
**base**
- [x] bump go to 1.16
- [x] bump rancher-machine to v0.15.0-rancher60

**use k8s api**
- [x] remove flags host, port, username, password
- [x] add flag `kubeconfig-content`

**compatible with Rancher 2.6 & RKE2**
- [x] add flag `cloud-config`
- [x] merge cloudinit userdata
- [x] use cloudinit secret
- [x] add flag `networkdata`
- [x] add flag `userdata`

**Related issues**
harvester/harvester#861

### CODE REVIEW

**Comment by futuretea** [comment]
need rancher bump rancher-machine first

File: harvester/client.go [hunk 1]
```diff
@@ -1,30 +1,252 @@
...
+	}
+	if namespace == "" {
+		namespace = defaultNamespace
+	}
+	return namespace, name, nil
+}
+
+func (d *Driver) getKubeConfigContent() ([]byte, error) {
+	if d.KubeConfigAddress != "" {
+		if err := Download(d.KubeConfigAddress, localKubeConfigSavePath); err != nil {
```
Comment (guangbochen) [inline harvester/client.go:110]: remove the `KubeConfigAddress`

File: harvester/create.go [hunk 2]
```diff
@@ -86,51 +86,65 @@ func (d *Driver) PreCreateCheck() error {
...
+				APIVersion: vm.APIVersion,
+				Kind:       vm.Kind,
+				Name:       vm.Name,
+				UID:        createdVM.UID,
+			},
+		}
+		if _, err = d.createSecret(cloudConfigSecret); err != nil {
+			return err
+		}
+	}
```
Comment (guangbochen) [inline harvester/create.go:145]: does it has an owner reference, make sure when the VM is deleted the related secret is also removed.

File: harvester/flags.go [hunk 3]
```diff
@@ -26,36 +23,29 @@ const (
...
+			Name:   "harvester-kubeconfig-content-base64",
+			Usage:  "harvester kubeconfig content base64",
 		},
 		mcnflag.StringFlag{
-			EnvVar: "HARVESTER_PASSWORD",
-			Name:   "harvester-password",
-			Usage:  "harvester password",
+			EnvVar: "HARVESTER_KUBECONFIG_ADDRESS",
+			Name:   "harvester-kubeconfig-address",
+			Usage:  "harvester kubeconfig address",
```
Comment (guangbochen) [inline harvester/flags.go:38]: ditto

File: harvester/flags.go [hunk 4]
```diff
@@ -26,36 +23,29 @@ const (
...
-			Value:  defaultInClusterHost,
-		},
-		mcnflag.IntFlag{
-			EnvVar: "HARVESTER_PORT",
-			Name:   "harvester-port",
-			Usage:  "harvester port",
-			Value:  defaultInClusterPort,
+			EnvVar: "HARVESTER_KUBECONFIG_CONTENT",
+			Name:   "harvester-kubeconfig-content",
+			Usage:  "harvester kubeconfig content",
```
Comment (guangbochen) [inline harvester/flags.go:28]: specify the usage with a more detailed description.

**Review by guangbochen — CHANGES_REQUESTED** [review]

File: harvester/create.go (hunk 2 above)
Comment (futuretea) [reply harvester/create.go:145]: yes

File: harvester/cloudinit.go [hunk 5]
```diff
@@ -1,9 +1,19 @@
...
+	"github.com/harvester/harvester/pkg/builder"
+	"github.com/imdario/mergo"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
 
 const (
-	userDataHeader            = `#cloud-config`
+	userDataHeader = `#cloud-config
+`
```
[the remark this answers is not in the document]
Comment (futuretea) [reply harvester/cloudinit.go:16]: this format is correct

File: harvester/create.go [hunk 6]
```diff
@@ -6,25 +6,28 @@ import (
...
 )
 
-func (d *Driver) PreCreateCheck() error {
-	c, err := d.getClient()
-	if err != nil {
-		return err
-	}
+const (
+	rootDiskName  = "disk-0"
+	interfaceName = "nic-0"
```
[the remark this answers is not in the document]
Comment (futuretea) [reply harvester/create.go:25]: The user does not need to care about these values

**Review by guangbochen — APPROVED** [review]
LGTM, can u squash the last commit, thanks

**Comment by futuretea** [comment]
> LGTM, can u squash the last commit, thanks

Updated

**Pull request merged**